contrib/twitchtv/twirp/twirp.go

```diff
@@ -25,6 +26,7 @@ type contextKey int
 
 const (
 	twirpErrorKey contextKey = iota
+	twirpSpanKey  contextKey = iota
```

Comment:
Yes, I'd agree. This is not how I'd typically manage context keys, but I was trying to follow the pattern I saw existing in this file.

Do you want me to update the existing `twirpErrorKey` as well, or only use this pattern for the new `twirpSpanKey`?

Reply:
I noticed that the existing use of `twirpErrorKey` is less defensive than my code, in that it casts without a check (which can panic.)

https://github.com/DataDog/dd-trace-go/blob/e7d03f2c9f20f6b31de1ae1635192e4f3109702e/contrib/twitchtv/twirp/twirp.go#L204
Can you let me know how you'd like to proceed with this, or feel free to push up a commit with your preferred patterns?

Reply:
Thanks for fixing this, including updating the `twirpErrorKey`.
The code actually *does* check during the cast, but uses the underscore to ignore the value of the check. That will avoid the panic and leave `err` as `nil` if the value is `nil` or the value cannot be converted to a `twirp.Error`.

In our case, this is fine, since we won't report an error that isn't a `twirp.Error`. If you think it is better, we can log if there *was* a value, but it wasn't a `twirp.Error`


Looking deeper into our context key scheme, I think that it doesn't work for multiple keys of the same type. Empty structs have weird properties, including that multiple distinctly-declared objects are allowed to have the same address.

As it is now, the keys will be identical and end up referencing the same value (see [my example](https://play.golang.org/p/8YF1Kzx3ROn))

Instead, we can just declare distinct types, which should always work when comparing. (see [my other example](https://play.golang.org/p/bujoPQzT8RT))

Reply:
>> https://github.com/DataDog/dd-trace-go/blob/e7d03f2c9f20f6b31de1ae1635192e4f3109702e/contrib/twitchtv/twirp/twirp.go#L204

> The code actually _does_ check during the cast, but uses the underscore to ignore the value of the check. That will avoid the panic and leave `err` as `nil` if the value is `nil` or the value cannot be converted to a `twirp.Error`.

I agree. However, the panic I had in mind was a missing context value. For instance:

```go
context.Background().Value("not there").(error)
```

(https://play.golang.org/p/V3PakMSrviG)

My preference is that a potential panic in 3rd party code is even worse than an unwanted error. 

I'll follow your lead if you have a different design in mind.

***

> Instead, we can just declare distinct types, which should always work when comparing. (see my other example)

:+1: Done.

Reply:
I agree, we *definitely* don't want that to panic. What I mean is that the `err, _ :=` check *does* make sure that even a `nil` won't panic. We just end up with `err == nil`, which is fine to pass to `tracer.WithError`. See here: https://play.golang.org/p/-JOm-SLbGsT